### PR TITLE
#160452449 - Search functionality

### PR DESCRIPTION
**Pull request by macphilips**
#### What does this PR do?
- Enable users to search for articles/tags and filter the search result

#### Description of Task to be completed?
- [x] - Consume search endpoint
- [x] - Add tests to all components created
- [x] - Improve responsive of the navbar when the search bar is opened.
- [x] - Fix issue with dropdown when a user clicks on any element except the drop-down menu

Other tasks
- Refactor Navbar component to make it global, so that it appears on all routes.

#### How should this be manually tested?

- Start the application using `npm start`
- Navigate to `localhost:3000`.
- click on the search icon on the navbar and input your query

#### Any background context you want to provide?
- None
#### What are the relevant pivotal tracker stories?
[#160452449](https://www.pivotaltracker.com/story/show/160452449)
#### Screenshots (if appropriate)
screen shot 2018-11-23 at 7 12 27 pm

#### Questions:

### CODE REVIEW

File: src/components/navbar/NavBarDropDown.jsx [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+export default class NavBarDropDown extends Component {
+  constructor(props) {
+    super(props);
+    this.ref = React.createRef();
+  }
+
+  // eslint-disable-next-line camelcase
+  UNSAFE_componentWillReceiveProps(newProps) {
+    // this.setState({ show: newProps.show });
```
Comment (blackshady) [inline src/components/navbar/NavBarDropDown.jsx]: Nice job on your PR, however, if you no longer need this code please remove it.

File: src/components/navbar/search/images/search-bar-icon.svg [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+<g>
+</g>
+<g>
+</g>
+<g>
+</g>
+<g>
+</g>
+<g>
+</g>
```
Comment (blackshady) [inline src/components/navbar/search/images/search-bar-icon.svg:43]: Please, why do you have multiple `g` tag on this file?

File: src/components/navbar/search/search.scss [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+.search-view {
+  height: 60px;
+  width: 600px;
+  padding: 0;
+  &.search-bar {
+    float: right;
+    height: 30px;
+    display: inline-block;
+    width: 300px;
+    //margin-top: 1.8rem;
```
Comment (blackshady) [inline src/components/navbar/search/search.scss]: Please remove the commented code if it is no longer needed.

File: src/components/navbar/search/search.scss [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+    .search-input {
+      height: 30px;
+      font-size: 1rem;
+      border: none;
+      font-weight: 400;
+      padding-right: 30px;
+      visibility: hidden;
+      transition: width .6s cubic-bezier(0.000, 0.795, 0.000, 1.000);
+      //transition: width .4s cubic-bezier(0.000, 0.795, 0.000, 1.000);
```
Comment (blackshady) [inline src/components/navbar/search/search.scss]: Please remove the commented code if it is no longer needed.

File: src/components/navbar/style.scss [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    display: none;
+    padding: 0;
+    top: 46px;
+    transform: translate(-50%, 0);
+    background: white;
+    min-width: 10px;
+    width: 280px;
+    z-index: 150;
+    //-webkit-box-shadow: 0 1px 2px rgba(0,0,0,.25), 0 0 1px rgba(0,0,0,.35);
+    //box-shadow: 0 1px 2px rgba(0,0,0,.25), 0 0 1px rgba(0,0,0,.35);
```
Comment (blackshady) [inline src/components/navbar/style.scss]: Please remove the commented code if it is no longer needed.

File: src/components/navbar/style.scss [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+    min-height: 150px;
+    border: 1px solid #ccc;
+    //border-top: 0;
```
Comment (blackshady) [inline src/components/navbar/style.scss]: Please remove the commented code if it is no longer needed.

File: src/tests/components/navbar/search.dropdown.test.js [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+  //     });
+  //   wrapper.find('#search_submit')
+  //     .simulate('click', { preventDefault });
+  //   wrapper.find('#search_submit')
+  //     .simulate('click', { preventDefault });
+  //   jest.runAllTimers();
+  //   expect(preventDefault)
+  //     .toBeCalled();
+  //   expect(props.onSubmit)
+  //     .toBeCalledWith({ q: props.query });
```
Comment (blackshady) [inline src/tests/components/navbar/search.dropdown.test.js]: Please remove the commented code if it is no longer needed.

**Review by olusoladavid — COMMENTED** [review]
Great job on this functionality @macphilips . I love the real-time search suggestions and categories that you implemented. However, I noticed that when a user searches for an item that does not match any tag or article in the database, you don't give the user any form of feedback to help them know that no search results were found for their search query and/or parameters. It would be nice to update the pull request with this requirement to provide a richer experience for the end user. Thanks.

**Comment by macphilips** [comment]
Hi @olusoladavid, thanks for the feedback. I have implemented the feedback, kindly review the changes. Cheers

**Pull request merged**